crypto/s2n_dhe.c

```diff
@@ -134,18 +134,20 @@ int s2n_check_all_dh_params(struct s2n_dh_params *dh_params)
     return S2N_SUCCESS;
 }
 
+DEFINE_POINTER_CLEANUP_FUNC(DH *, DH_free);
 int s2n_pkcs3_to_dh_params(struct s2n_dh_params *dh_params, struct s2n_blob *pkcs3)
 {
     POSIX_ENSURE_REF(dh_params);
     POSIX_PRECONDITION(s2n_blob_validate(pkcs3));
 
     uint8_t *original_ptr = pkcs3->data;
-    dh_params->dh = d2i_DHparams(NULL, (const unsigned char **) (void *) &pkcs3->data, pkcs3->size);
+    DEFER_CLEANUP(DH *dh_val = d2i_DHparams(NULL, (const unsigned char **) (void *) &pkcs3->data, pkcs3->size), DH_free_pointer);
+    dh_params->dh = dh_val;
     POSIX_GUARD(s2n_check_p_g_dh_params(dh_params));
-    if (pkcs3->data && (pkcs3->data - original_ptr != pkcs3->size)) {
-        DH_free(dh_params->dh);
-        POSIX_BAIL(S2N_ERR_INVALID_PKCS3);
-    }
+    POSIX_ENSURE_REF(pkcs3->data);
+    POSIX_ENSURE_GT(pkcs3->data, original_ptr);
+    POSIX_ENSURE_EQ((uint32_t) (pkcs3->data - original_ptr), pkcs3->size);
+
     pkcs3->data = original_ptr;
 
     /* Require at least 2048 bits for the DH size */
```

Comment:
I can't highlight the check here, but you should replace the check on line 152/154 just like you replaced the one on line 145. If you don't, and you also have already set dh_params to dh_val, you're going to double free here.

Are there any tests for this method?

Reply:
I couldn't find anything in particular for this method, but there are cases where this method has been used in other tests. Test like-

- https://github.com/aws/s2n-tls/blob/90dd13aa7d9b4048d5f2d0ef44402222947407a3/tests/unit/s2n_pem_rsa_dhe_test.c#L131

Reply:
Notice those are all EXPECT_SUCCESS :) The double free would only have been caught by a test that triggered the right failure. We probably shouldn't make risky memory management choices here unless we're going to write some much more thorough tests.